File: 프로그래머스/3/12927. 야근 지수/야근 지수.py
def solution(n: int, works: list):
    works.sort(reverse=True)

    while n > 0:
        for i, _ in enumerate(works):
            works[i] -= 1
            n -= 1
            
            if n == 0:
                return sum(x ** 2 for x in works if x > 0)
            if i < len(works) - 1 and works[i] >= works[i + 1]:
                break
                
    return 0

File: 프로그래머스/3/12927. 야근 지수/야근 지수.java
import java.util.*;

class Solution {

    public long solution(int n, int[] works) {
        PriorityQueue<Integer> pq = new PriorityQueue<>(Collections.reverseOrder());
        for (int w : works) if (w > 0) pq.offer(w);

        while (n-- > 0 && !pq.isEmpty()) {
            int top = pq.poll();
            if (top == 0) break;
            pq.offer(top - 1);
        }

        long sum = 0;
        for (int v : pq) sum += (long) v * v;
        return sum;
    }
}